CARDS-2372: Add support for data migrators

acrowthe opened the pull request:
Add a data migrator that tracks the previous version and checks if any created data migrators need to run.

Also includes CARDS-2336.

To test:
- build and run an older version of cards:
- `git checkout cards-0.9.15`
- `mvn clean install -P docker`
- `cd compose-cluster`
- `python3 generate_compose_yaml.py --dev_docker_image  --composum --cards_project cards4heracles --mongo_singular`
- `docker-compose build && docker-compose up`
- Create a few CPET Interpretation forms, fill in the Protocol question with a few Bike value and a few other values
- Stop docker
- `cd ..`
- `git checkout CARDS-2372`
- `mvn clean install -P docker`
- `cd compose-cluster`
- `python3 generate_compose_yaml.py --dev_docker_image  --composum --cards_project cards4heracles --mongo_singular`
- `docker-compose build && docker-compose up`
- Check that the previous Bike answers are now Bike 50 rpm and the previous other answers are unchanged
- Check that `localhost:8080/libs/cards/conf/PrevVersion.json` now has `0.9.0.SNAPSHOT` stored

acrowthe committed:
CARDS-2372: Add support for data migrators
- Code implemented, enountering issues with race conditions:
  - DataMigratorManager sometimes runs before UpdatePreviousVersionMigrator is loaded
  - DataMigratorManager sometimes runs before Version node is set on new install

acrowthe committed:
CARDS-2372: Add Support for data migrators
Fix race conditions:
- Pull current version from bundle context rather than node
- Run any needed migrators that are started after DataMigratorManager is run

acrowthe commented:
Needs investigation:
- The formatting of the current version as pulled from the bundle context does not match the version that is being stored in our database. This was observed in dev, I am unsure if the versions in prod will match as they do not include `-SNAPSHOT`.

Version string currently being stored:
![image](https://github.com/data-team-uhn/cards/assets/20056751/ea085973-eb65-48f7-b37e-e7435ee527a7)

Version string retrieved from the bundle and accessible when the migrators are run:
![image](https://github.com/data-team-uhn/cards/assets/20056751/52885256-6171-4886-a33f-efaa16ce30d8)

sdumitriu commented on modules/data-model/migrators/src/main/java/io/uhndata/cards/migrators/spi/DataMigrator.java:
```diff
@@ -0,0 +1,63 @@
+
+    /**
+     * If this migrator should run based on what version of CARDS was previously run.
+     * @param previousVersion The version of CARDs that was run previously
```
```suggestion
     * @param previousVersion The version of CARDs that was run previously, {@code null} if this is the first run
```

sdumitriu commented on modules/data-model/migrators/src/main/java/io/uhndata/cards/migrators/spi/DataMigrator.java:
```diff
@@ -0,0 +1,63 @@
+    boolean shouldRun(String previousVersion, String currentVersion, Session session);
+
+    /**
+     * Change anything that needs to be changed to upgrade to from the previous version of CARDS.
```
```suggestion
     * Change anything that needs to be changed to upgrade from the previous version of CARDS.
```

sdumitriu commented on modules/data-model/migrators/src/main/java/io/uhndata/cards/migrators/spi/DataMigrator.java:
```diff
@@ -0,0 +1,63 @@
+
+    /**
+     * Change anything that needs to be changed to upgrade to from the previous version of CARDS.
+     * @param previousVersion The version of CARDs that was run previously
```
```suggestion
     * @param previousVersion The version of CARDs that was run previously, may be {@code null}
```

sdumitriu commented on modules/data-model/migrators/src/main/java/io/uhndata/cards/migrators/internal/DataMigratorManager.java:
```diff
@@ -0,0 +1,128 @@
+        }
+    }
+
+    protected void migratorRemoved(final DataMigrator migrator)
```
I think we can skip the `unbind` setting completely.

sdumitriu replied:
Turns out it's needed, OSGi complains without it.

sdumitriu commented on modules/data-model/migrators/src/main/java/io/uhndata/cards/migrators/internal/DataMigratorManager.java:
```diff
@@ -0,0 +1,128 @@
+        final Map<String, Object> parameters =
+            Collections.singletonMap(ResourceResolverFactory.SUBSERVICE, "dataMigratorManager");
+        try (ResourceResolver resolver = this.resolverFactory.getServiceResourceResolver(parameters);) {
+            final Session session = resolver.adaptTo(Session.class);
```
The session should be pushed to our resolver provider, otherwise FormUtils and others may not work.

sdumitriu commented on modules/data-model/migrators/src/main/java/io/uhndata/cards/migrators/internal/DataMigratorManager.java:
```diff
@@ -0,0 +1,128 @@
+
+            // Run any required migrators
+            for (DataMigrator migrator: sortedMigrators) {
+                runMigrator(migrator, session);
```
Missing the `shouldRun` check.

sdumitriu replied:
Nevermind, it's in the `runMigrator` method itself.

sdumitriu commented on modules/data-model/migrators/src/main/java/io/uhndata/cards/migrators/spi/DataMigrator.java:
```diff
@@ -0,0 +1,63 @@
+
+import javax.jcr.Session;
+
+public interface DataMigrator extends Comparable<DataMigrator>
```
Needs some javadoc.

sdumitriu committed:
CARDS-2372: Add Support for data migrators
Make the ResourceResolver available in other components

sdumitriu committed:
[misc] Applied codestyle

sdumitriu committed:
CARDS-2372: Add Support for data migrators
Switch from simple string to actual Version objects for passing the previous and current version to migrators

sdumitriu committed:
CARDS-2336: Migrate CPET Interpretation Protocol's "Bike" answer to "Bike 50 rpm"

sdumitriu committed:
CARDS-2372: Add Support for data migrators
Javadoc improvements

sashaandjic commented:
![image](https://github.com/data-team-uhn/cards/assets/23243244/22852b5c-413e-44cd-bc64-44c04a8806ed)
![image](https://github.com/data-team-uhn/cards/assets/23243244/2c0e2def-9964-4e84-bec9-b535da2ee4a8)

sashaandjic reviewed (approved):
data was migrated from `Bike` to `Bike 50rpm`

![image](https://github.com/data-team-uhn/cards/assets/23243244/eaade53d-3b29-4c05-92cd-a8ed8de791b2)

`http://localhost:8080/libs/cards/conf/PrevVersion.json` has `0.9.0.SNAPSHOT` as expected
![image](https://github.com/data-team-uhn/cards/assets/23243244/19d51319-dcf3-4e83-97bc-91e577ab1d22)

sashaandjic commented:
The old option is still available:
![image](https://github.com/data-team-uhn/cards/assets/23243244/1d2d2fd6-6663-4a7c-894f-42542c65d66c)

sdumitriu commented:
> The old option `Bike` is still available, it should have been removed. ![image](https://user-images.githubusercontent.com/23243244/275130529-1d2d2fd6-6663-4a7c-894f-42542c65d66c.png)

Not quite, this is "expected", questionnaires are not updated when upgrading the version. It's a know issue.

sashaandjic commented:
> Not quite, this is "expected", questionnaires are not updated when upgrading the version. It's a know issue.

OK. I will mark this PR as `tested` then.

sashaandjic reviewed (approved):
tested

sdumitriu committed:
CARDS-2372: Add Support for data migrators
Add a default priority of 0
Also sort migrators of equal priority by their name

sdumitriu committed:
CARDS-2372: Add Support for data migrators
Simplify service session configuration

sdumitriu committed:
CARDS-2372: Add Support for data migrators
Fixed NPE when there was no previous version
Better detection of upgrades
